packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts

```diff
@@ -66,13 +86,118 @@ export default class PaymentNetworkERC20ProxyContract implements Types.IPaymentN
+    const paymentNetworkId = ExtensionTypes.ID.PAYMENT_NETWORK_ERC20_PROXY_CONTRACT;
+    const paymentNetwork = request.extensions[paymentNetworkId];
+
+    if (!paymentNetwork) {
+      throw new Error(`The request do not have the extension : ${paymentNetworkId}`);
+    }
+
+    const paymentAddress = paymentNetwork.values.paymentAddress;
+    const refundAddress = paymentNetwork.values.refundAddress;
+    const salt = paymentNetwork.values.salt;
+
+    let payments: Types.IBalanceWithEvents = { balance: '0', events: [] };
+    if (paymentAddress) {
+      payments = await this.extractBalanceAndEvents(
+        request,
+        salt,
+        paymentAddress,
+        Types.EVENTS_NAMES.PAYMENT,
+        paymentNetwork.version,
+      );
+    }
+
+    let refunds: Types.IBalanceWithEvents = { balance: '0', events: [] };
+    if (refundAddress) {
+      refunds = await this.extractBalanceAndEvents(
+        request,
+        salt,
+        refundAddress,
+        Types.EVENTS_NAMES.REFUND,
+        paymentNetwork.version,
+      );
+    }
+
+    const balance: string = new bigNumber(payments.balance || 0)
+      .sub(new bigNumber(refunds.balance || 0))
+      .toString();
+
+    const events: Types.ERC20PaymentNetworkEvent[] = [...payments.events, ...refunds.events].sort(
+      (a, b) => (a.timestamp || 0) - (b.timestamp || 0),
+    );
+
+    return {
+      balance,
+      events,
+    };
+  }
+
+  /**
+   * Extracts the balance and events of an address
+   *
+   * @private
+   * @param address Address to check
+   * @param eventName Indicate if it is an address for payment or refund
+   * @param network The id of network we want to check
+   * @param tokenContractAddress the address of the token contract
+   * @returns The balance
+   */
+  private async extractBalanceAndEvents(
+    request: RequestLogicTypes.IRequest,
+    salt: string,
+    toAddress: string,
+    eventName: Types.EVENTS_NAMES,
+    paymentNetworkVersion: string,
+  ): Promise<Types.IBalanceWithEvents> {
+    const network = request.currency.network;
+
+    if (!network) {
+      throw new Error(`Payment network not supported by ERC20 payment detection`);
+    }
+
+    if (!PROXY_CONTRACT_ADDRESS_BY_VERSION_BY_NETWORK[paymentNetworkVersion]) {
+      throw new Error(`Payment network version not supported: ${paymentNetworkVersion}`);
+    }
+
+    const proxyContractAddress: string | undefined =
+      PROXY_CONTRACT_ADDRESS_BY_VERSION_BY_NETWORK[paymentNetworkVersion][network];
+
+    if (!proxyContractAddress) {
+      throw new Error(
+        `Network not supported for this payment network: ${request.currency.network}`,
+      );
+    }
+
+    const paymentReference = PaymentReferenceCalculator.calculate(
+      request.requestId,
+      salt,
+      toAddress,
+    );
+
+    const infoRetriever = new ProxyInfoRetriever(
+      paymentReference,
+      proxyContractAddress,
```

Comment:
Shouldn't this value be retrieved inside ProxyInfoRetriever since it's the only place where it's used

Reply:
It is dependent on the network and the version of the payment network, I would rather keep this outisde `ProxyInfoRetriever` that takes care only about reading and parsing logs